README
ibus.el README

----------------------------------------------------------------------
0. Introduction
----------------------------------------------------------------------

IBus is a new input method framework under active development which
is designed to overcome the limitations of SCIM.

IBus uses D-Bus protocol for communication between the ibus-daemon
and clients (engines, panel, config tools). Since the components
run in separate processes there is enhanced modularity and stability.
Client processes can be loaded, started and stopped independently.
The input method engines and clients can be written in any language
with a dbus binding.

ibus.el is a IBus client for GNU Emacs. This program allows users
on-the-spot style input with IBus. The input statuses are individually
kept for each buffer, and prefix-keys such as C-x and C-c can be used
even if IBus is active. So you can input various languages fast and
comfortably by using it.

This program is *not* a part of IBus.

----------------------------------------------------------------------
1. License
----------------------------------------------------------------------

GNU General Public License version 3

----------------------------------------------------------------------
2. Requirements
----------------------------------------------------------------------

emacs22 | emacs23 | emacs-snapshot
python (>= 2.5)
ibus (>= 1.2)
python-xlib

----------------------------------------------------------------------
3. Installation
----------------------------------------------------------------------

i) Common settings

Save this file as ibus.el and byte-compile in a directory that is
listed in load-path, and also save ibus-el-agent somewhere in
your system.

Put the following in your .emacs file:

 (require 'ibus)
 (add-hook 'after-init-hook 'ibus-mode-on)

If ibus.el and ibus-el-agent are saved in different directories,
add a setting to the above as follows:

 (setq ibus-agent-file-name "/PATH/TO/ibus-el-agent")

To disable XIM in Emacs, put the following in ~/.Xresources:

 Emacs*useXIM: false

and restart X session or execute a shell command:

 xrdb ~/.Xresources

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

ii) Using ibus-anthy

To toggle half-width eisu mode by C-j, add the following to .emacs:

 (ibus-define-common-key ?\C-j t)

To use kana input method with jp106 keyboard, you can enable kana
onbiki key as follows:

 (setq ibus-use-kana-onbiki-key t)

If you use thumb shift input method, you have to specify the
simultaneous pressing time as:

 (setq ibus-ibus-simultaneous-pressing-time 0.1)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

iii) Using ibus-chewing

Please set input style to "in application window" in ibus-chewing's
configuration dialog.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

iv) Example of settings in .emacs

 (require 'ibus)
 ;; Turn on ibus-mode automatically after loading .emacs
 (add-hook 'after-init-hook 'ibus-mode-on)
 ;; Use C-SPC for Set Mark command
 (ibus-define-common-key ?\C-\s nil)
 ;; Use C-/ for Undo command
 (ibus-define-common-key ?\C-/ nil)
 ;; Change cursor color depending on IBus status
 (setq ibus-cursor-color '("red" "blue" "limegreen"))

----------------------------------------------------------------------
4. Web sites
----------------------------------------------------------------------

Home page (in Japanese):

 http://www11.atwiki.jp/s-irie/pages/21.html

Wiki:

 http://www.emacswiki.org/emacs/IBusMode

Development:

 https://launchpad.net/ibus.el

Bug tracker:

 https://bugs.launchpad.net/ibus.el

----------------------------------------------------------------------
5. Author
----------------------------------------------------------------------

IRIE Shinsuke <irieshinsuke@yahoo.co.jp>


ChangeLog
	* Version0.1.0
	* First public release
	* Added Multi-display support
	* Added Japanese kana typing method support
	* Added Japanese thumb shift typing method support
	* Fixed a lot of bugs
2010-05-09  IRIE Shinsuke <irieshinsuke@yahoo.co.jp>
	* Version0.0.2
	* Imported many functions from scim-bridge.el
2010-04-12  IRIE Shinsuke <irieshinsuke@yahoo.co.jp>
	* Version0.0.1
	* Initial experimental version


make-dist.sh
#!/bin/bash
# Copyright (c) 2010 and onwards, S. Irie
# This program is distributed under the MIT Licence.

EL_DEV="ibus-dev.el"
EL_MAIN="ibus.el"
DBGREGEXP="^;*\t* *(ibus-log\b"

FILES=("ibus-el-agent" "README" "doc" "debian")

PACKAGE="ibus-el"
VERSION=$(sed -n 's/^(defconst ibus-mode-version "\(.*\)")$/\1/p' $EL_DEV)


# Command line options =================

DCH="no"
DEBUILD="no"
BLDTYPE="deb"
NATIVE="yes"
BLDFLAGS=""

while getopts :hcbSpnu OPTION
do
    case $OPTION in
	h)
	    echo "usage: $0 [OPTION]..."
	    echo "  -c  edit debian/changelog"
	    echo "  -b  build deb package"
	    echo "  -S  build source package"
	    echo "  -p  build deb package using pbuilder"
	    echo "  -n  build deb package as a non-native one"
	    echo "  -u  add -us -uc options to debuild"
	    echo "  -h  display this help and exit"
	    exit
	    ;;
	c)
	    DCH="yes"
	    ;;
	b)
	    DEBUILD="yes"
	    ;;
	S)
	    if [ "$BLDTYPE" == "pbuild" ]; then
		echo "$0: options -p and -S can't be used together." >&2
		exit 1
	    fi

	    DEBUILD="yes"
	    BLDTYPE="src"
	    ;;
	p)
	    if [ "$BLDTYPE" == "src" ]; then
		echo "$0: options -S and -p can't be used together." >&2
		exit 1
	    fi

	    DEBUILD="yes"
	    BLDTYPE="pbuild"
	    ;;
	n)
	    NATIVE="no"
	    ;;
	u)
	    BLDFLAGS="${BLDFLAGS} -us -uc"
	    ;;
	?)
	    echo "$0: invalid option -- $OPTARG" >&2
	    echo "Try \`$0 -h' for more information." >&2
	    exit 1
	    ;;
    esac
done


# Paths ================================

WORKDIR="v${VERSION}"
ARCHDIR="${WORKDIR}/${PACKAGE}-${VERSION}"
ARCHFILE="${ARCHDIR}.tar.gz"
ARDEBDIR="${ARCHDIR}/debian"
ORIGARCH="${WORKDIR}/${PACKAGE}_${VERSION}.orig.tar.gz"
CURRDIR="$PWD"


# Locate files =========================

rm -rfv $WORKDIR
mkdir -pv $ARCHDIR

if [ "$DBGREGEXP" == "" ]; then
    cp -pv $EL_DEV ${ARCHDIR}/${EL_MAIN}
else
    echo "comment out debug codes in $EL_DEV -> ${ARCHDIR}/${EL_MAIN}"
    sed "s/\(${DBGREGEXP}\)/;#\1/" $EL_DEV > ${ARCHDIR}/${EL_MAIN}
fi

for i in ${FILES[@]}; do
    cp -rpv $i $ARCHDIR
done

for i in $(find $ARCHDIR -regex '.*\(~\|\.~[0-9]+~\)'); do
    rm -fv $i      # Delete backup files
done


# Edit changelog =======================

if [ "$DCH" == "yes" ]; then
    cd $ARDEBDIR
    dch -v ${VERSION}
    cd $CURRDIR

    cp -pv --backup=t ${ARDEBDIR}/changelog debian/changelog
fi


# Create archive =======================

echo "create archive $ARCHFILE"
cd $WORKDIR
tar cvzf $(basename $ARCHFILE) $(basename $ARCHDIR)
cd $CURRDIR


# Build deb package  ===================

if [ ! "$DEBUILD" == "yes" ]; then
    exit 0
fi

if [ "$NATIVE" == "no" ]; then
    cp -pv $ARCHFILE $ORIGARCH
fi

cd $ARCHDIR

case $BLDTYPE in
    deb)
	debuild ${BLDFLAGS}
	;;
    src)
	debuild -S ${BLDFLAGS}
	;;
    pbuild)
	debuild -S ${BLDFLAGS} && sudo pbuilder build ../*.dsc
	;;
esac

cd $CURRDIR


ibus-agent.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# Version 0.0.1.1

import sys
import glib
import ibus

# D-Bus

bus = ibus.Bus()
path = bus.create_input_context("IBusAgentIMContext")

# Miscellaneous functions

def lisp_boolean(boolean):
    return "t" if boolean else "nil"

# Class definition

class IBusAgentIMContext(ibus.InputContext):

    def __init__(self):
        super(IBusAgentIMContext, self).__init__(bus, path, True)

        self.id_no = 0
        self.lookup_table = None

        self.connect('commit-text', self.__commit_text_cb)
        self.connect('update-preedit-text', self.__update_preedit_text_cb)
        self.connect('show-preedit-text', self.__show_preedit_text_cb)
        self.connect('hide-preedit-text', self.__hide_preedit_text_cb)
        self.connect('update-auxiliary-text', self.__update_auxiliary_text_cb)
        self.connect('show-auxiliary-text', self.__show_auxiliary_text_cb)
        self.connect('hide-auxiliary-text', self.__hide_auxiliary_text_cb)
        self.connect('update-lookup-table', self.__update_lookup_table_cb)
        self.connect('show-lookup-table', self.__show_lookup_table_cb)
        self.connect('hide-lookup-table', self.__hide_lookup_table_cb)
        self.connect('page-up-lookup-table', self.__page_up_lookup_table_cb)
        self.connect('page-down-lookup-table', self.__page_down_lookup_table_cb)
        self.connect('cursor-up-lookup-table', self.__cursor_up_lookup_table_cb)
        self.connect('cursor-down-lookup-table', self.__cursor_down_lookup_table_cb)
        self.connect('enabled', self.__enabled_cb)
        self.connect('disabled', self.__disabled_cb)

    # Callbacks

    def __commit_text_cb(self, ic, text):
        print '(ibus-commit-text-cb %d "%s")'%(ic.id_no, text.text.encode("utf-8"))

    def __update_preedit_text_cb(self, ic, text, cursor_pos, visible):
        attrs = ['%s %d %d %d'%
                 (["nil", "'underline", "'foreground", "'background"][attr.type],
                  attr.value & 0xffffff, attr.start_index, attr.end_index)
                 for attr in text.attributes]
        print '(ibus-update-preedit-text-cb %d "%s" %d %s %s)'% \
            (ic.id_no, text.text.encode("utf-8"),
             cursor_pos, lisp_boolean(visible), ' '.join(attrs))

    def __show_preedit_text_cb(self, ic):
        print '(ibus-show-preedit-text-cb %d)'%(ic.id_no)

    def __hide_preedit_text_cb(self, ic):
        print '(ibus-hide-preedit-text-cb %d)'%(ic.id_no)

    def __update_auxiliary_text_cb(self, ic, text, visible):
        print '(ibus-update-auxiliary-text-cb %d "%s" %s)'% \
            (ic.id_no, text.text.encode("utf-8"), lisp_boolean(visible))

    def __show_auxiliary_text_cb(self, ic):
        print '(ibus-show-auxiliary-text-cb %d)'%(ic.id_no)

    def __hide_auxiliary_text_cb(self, ic):
        print '(ibus-hide-auxiliary-text-cb %d)'%(ic.id_no)

    def __update_lookup_table_cb(self, ic, lookup_table, visible):
        ic.lookup_table = lookup_table
        if visible:
            self.__show_lookup_table_cb(ic)
        else:
            self.__hide_lookup_table_cb(ic)

    def __show_lookup_table_cb(self, ic):
        print "(ibus-show-lookup-table-cb %d '(%s))"% \
            (ic.id_no,
             " ".join(map(lambda item : '"%s"'%item.text,
                          ic.lookup_table.get_candidates_in_current_page())
                      ).encode("utf-8"))

    def __hide_lookup_table_cb(self, ic):
        print '(ibus-hide-lookup-table-cb %d)'%(ic.id_no)

    def __page_up_lookup_table_cb(self, ic):
        pass

    def __page_down_lookup_table_cb(self, ic):
        pass

    def __cursor_up_lookup_table_cb(self, ic):
        pass

    def __cursor_down_lookup_table_cb(self, ic):
        pass

    def __enabled_cb(self, ic):
        print '(set-cursor-color "red")'

    def __disabled_cb(self, ic):
        print "(set-cursor-color (frame-parameter nil 'foreground-color))"


# Process methods from client

imcontexts = []

def create_imcontext():
    ic = IBusAgentIMContext()
    imcontexts.append(ic)
    ic.id_no = len(imcontexts)-1
    ic.set_capabilities(7)
    print "(ibus-create-imcontext-cb %d)"%ic.id_no

def destroy_imcontext(id_no):
    imcontexts[id_no].destroy()
    imcontexts[id_no] = None

def process_key_event(id_no, keyval, keycode, modifier_mask):
    handled = imcontexts[id_no].process_key_event(keyval, keycode, modifier_mask)
    print "(ibus-process-key-event-cb %d %s)"%(id_no, lisp_boolean(handled))

# Main loop

mainloop = None

def quit_ibus():
    mainloop.quit()

def stdin_cb(fd, condition):
    try:
        exec sys.stdin.readline()
    except:
        import traceback
        print '(error "%s")'%traceback.format_exc().replace('"', '\\"')
    return True

def main():
    global mainloop
    mainloop = glib.MainLoop()
    glib.io_add_watch(0, glib.IO_IN, stdin_cb)
    while True:
        try:
            mainloop.run()
        except:
            import traceback
            print '(error "%s")'%traceback.format_exc().replace('"', '\\"')
        else:
            break
    for ic in imcontexts:
        if ic:
            ic.destroy()
    print '(message "Bye.")'

if __name__ == "__main__":
    main()


ibus_el_agent.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# ibus_el_agent.py --- helper program of IBus client for GNU Emacs
# Copyright (c) 2009 and onwards, S. Irie

# Author: S. Irie
# Maintainer: S. Irie
# Version: 0.0.2.57

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Commentary:

# IBus is a new input method framework under active development
# which is designed to overcome the limitations of SCIM.

# IBus uses D-Bus protocol for communication between the ibus-daemon
# and clients (engines, panel, config tools). Since the components
# run in separate processes there is enhanced modularity and stability.
# Client processes can be loaded, started and stopped independently.
# IBus supports Gtk2 and XIM, and has input method engines for anthy,
# chewing, hangul, m17n, pinyin, rawcode, and large tables. Engines
# and clients can be written in any language with a dbus binding.

# This program is IBus client for GNU Emacs. It is, however,
# not part of official IBus project.

# ChangeLog:

# 2010-05-09
#        * Version 0.0.2
# 2010-04-12
#        * Version 0.0.1
#        * Initial experimental version

# Code:

import sys
import glib
import re

import ibus
from ibus import modifier

try:
    import Xlib.display
except ImportError:
    print "python-xlib library is required"
    exit(1)

display = Xlib.display.Display()

########################################################################
# Connect to IBus daemon
########################################################################

try:
    bus = ibus.Bus()
except TypeError:
    import os
    import time
    if os.spawnlp(os.P_WAIT, "ibus-daemon", "ibus-daemon", "-d") == 0:
        for i in range(10):
            time.sleep(0.5)
            try:
                bus = ibus.Bus()
                break
            except TypeError:
                pass
        else:
            print '(error "Failed to connect with ibus-daemon")'
            exit(1)
    else:
        print '(error "Failed to launch ibus-daemon")'
        exit(1)

########################################################################
# Miscellaneous functions
########################################################################

def lisp_boolean(boolean):
    return "t" if boolean else "nil"

escape_regexp = re.compile(ur'(["\\])')

def escape_string(string):
    return escape_regexp.sub(ur'\\\1', string)

########################################################################
# Input Context
########################################################################

class IBusELInputContext(ibus.InputContext):

    def __init__(self, bus):
        self.__bus = bus
        self.__path = bus.create_input_context("IBusELInputContext")
        super(IBusELInputContext, self).__init__(bus, self.__path, True)

        self.id_no = 0
        self.lookup_table = None

        self.connect('commit-text', commit_text_cb)
        self.connect('update-preedit-text', update_preedit_text_cb)
        self.connect('show-preedit-text', show_preedit_text_cb)
        self.connect('hide-preedit-text', hide_preedit_text_cb)
        self.connect('update-auxiliary-text', update_auxiliary_text_cb)
        self.connect('show-auxiliary-text', show_auxiliary_text_cb)
        self.connect('hide-auxiliary-text', hide_auxiliary_text_cb)
        self.connect('update-lookup-table', update_lookup_table_cb)
        self.connect('show-lookup-table', show_lookup_table_cb)
        self.connect('hide-lookup-table', hide_lookup_table_cb)
        self.connect('page-up-lookup-table', page_up_lookup_table_cb)
        self.connect('page-down-lookup-table', page_down_lookup_table_cb)
        self.connect('cursor-up-lookup-table', cursor_up_lookup_table_cb)
        self.connect('cursor-down-lookup-table', cursor_down_lookup_table_cb)
        self.connect('enabled', enabled_cb)
        self.connect('disabled', disabled_cb)
        try:
            self.connect('forward-key-event', forward_key_event_cb)
        except TypeError:
            pass
        try:
            self.connect('delete-surrounding-text', delete_surrounding_text_cb)
        except TypeError:
            pass

########################################################################
# Callbacks
########################################################################

def commit_text_cb(ic, text):
    print '(ibus-commit-text-cb %d "%s")'% \
        (ic.id_no, escape_string(text.text).encode("utf-8"))

def update_preedit_text_cb(ic, text, cursor_pos, visible):
    attrs = ['%s %d %d %d'%
             (["nil", "'underline", "'foreground", "'background"][attr.type],
              attr.value & 0xffffff, attr.start_index, attr.end_index)
             for attr in text.attributes]
    print '(ibus-update-preedit-text-cb %d "%s" %d %s %s)'% \
        (ic.id_no, escape_string(text.text).encode("utf-8"),
         cursor_pos, lisp_boolean(visible), ' '.join(attrs))

def show_preedit_text_cb(ic):
    print '(ibus-show-preedit-text-cb %d)'%(ic.id_no)

def hide_preedit_text_cb(ic):
    print '(ibus-hide-preedit-text-cb %d)'%(ic.id_no)

def update_auxiliary_text_cb(ic, text, visible):
    print '(ibus-update-auxiliary-text-cb %d "%s" %s)'% \
        (ic.id_no, escape_string(text.text).encode("utf-8"),
         lisp_boolean(visible))

def show_auxiliary_text_cb(ic):
    print '(ibus-show-auxiliary-text-cb %d)'%(ic.id_no)

def hide_auxiliary_text_cb(ic):
    print '(ibus-hide-auxiliary-text-cb %d)'%(ic.id_no)

def update_lookup_table_cb(ic, lookup_table, visible):
    ic.lookup_table = lookup_table
    if visible:
        self.__show_lookup_table_cb(ic)
    else:
        self.__hide_lookup_table_cb(ic)

def show_lookup_table_cb(ic):
    print "(ibus-show-lookup-table-cb %d '(%s) %s)"% \
        (ic.id_no, escape_string(
            " ".join(map(lambda item : '"%s"'%item.text,
                         ic.lookup_table.get_candidates_in_current_page())
                     )).encode("utf-8"),
         ic.lookup_table.get_cursor_pos_in_current_page())

def hide_lookup_table_cb(ic):
    print '(ibus-hide-lookup-table-cb %d)'%(ic.id_no)

def page_up_lookup_table_cb(ic):
    print '(ibus-log "page up lookup table")'

def page_down_lookup_table_cb(ic):
    print '(ibus-log "page down lookup table")'

def cursor_up_lookup_table_cb(ic):
    print '(ibus-log "cursor up lookup table")'

def cursor_down_lookup_table_cb(ic):
    print '(ibus-log "cursor down lookup table")'

def enabled_cb(ic):
    print '(ibus-status-changed-cb %d "%s")'%(ic.id_no, ic.get_engine().name)

def disabled_cb(ic):
    print '(ibus-status-changed-cb %d nil)'%ic.id_no

def forward_key_event_cb(ic, keyval, keycode, modifiers):
    print '(ibus-forward-key-event-cb %d %d %d %s)'% \
        (ic.id_no, keyval, modifiers & ~modifier.RELEASE_MASK,
         lisp_boolean(modifiers & modifier.RELEASE_MASK == 0))

def delete_surrounding_text_cb(ic, offset, n_chars):
    print '(ibus-delete-surrounding-text-cb %d %d %d)'% \
        (ic.id_no, offset, n_chars)

########################################################################
# Process methods from client
########################################################################

imcontexts = []

def create_imcontext():
    ic = IBusELInputContext(bus)
    try:
        ic.id_no = imcontexts.index(None)
        imcontexts[ic.id_no] = ic
    except ValueError:
        ic.id_no = len(imcontexts)
        imcontexts.append(ic)
    ic.set_capabilities(int('101001',2))
    print '(ibus-create-imcontext-cb %d)'%ic.id_no

def destroy_imcontext(id_no):
    imcontexts[id_no].destroy()
    if id_no == len(imcontexts) - 1:
        imcontexts.pop()
    else:
        imcontexts[id_no] = None

def process_key_event(id_no, keyval, modmask, backslash, pressed):
    if backslash:
        keycode = (display.keysym_to_keycode(backslash), 0)
    else:
        keycode = display.keysym_to_keycodes(keyval)[0]
        if keyval < 0x100 and keycode[1] & 1 != 0:
            modmask |= modifier.SHIFT_MASK
    if not pressed:
        modmask |= modifier.RELEASE_MASK
    handled = imcontexts[id_no].process_key_event(keyval, keycode[0] - 8, modmask)
    print '(ibus-process-key-event-cb %d %s)'%(id_no, lisp_boolean(handled))

def set_cursor_location(id_no, x, y, w, h):
    imcontexts[id_no].set_cursor_location(x, y, w, h)

def focus_in(id_no):
    imcontexts[id_no].focus_in()

def focus_out(id_no):
    imcontexts[id_no].focus_out()
    print '()' # Dummy response

def reset(id_no):
    imcontexts[id_no].reset()

def enable(id_no):
    imcontexts[id_no].enable()

def disable(id_no):
    imcontexts[id_no].disable()

def set_engine(id_no, name):
    for engine in bus.list_active_engines():
        if name == '%s'%engine.name:
            imcontexts[id_no].set_engine(engine)
            break
    else:
        enable(id_no)

########################################################################
# Experiment
########################################################################

def list_active_engines():
    print '(ibus-log "active engines: %s")'% \
        ' '.join('%s'%i.name for i in bus.list_active_engines())

########################################################################
# Main loop
########################################################################

class IBusModeMainLoop(glib.MainLoop):

    def __init__(self, bus):
        super(IBusModeMainLoop, self).__init__()
        bus.connect("disconnected", self.__disconnected_cb)

    def __disconnected_cb(self, *args):
        print '(ibus-log "disconnected")'
        exit()

    def __start_cb(self):
        print '(setq started t)'
        print 'Agent successfully started for display "%s"'% \
            display.get_display_name()
        return False

    def __stdin_cb(self, fd, condition):
        try:
            exec sys.stdin.readline()
        except:
            import traceback
            print '(error "%s")'%traceback.format_exc().replace('"', '\\"')
        return True

    def run(self):
        glib.idle_add(self.__start_cb)
        glib.io_add_watch(0, glib.IO_IN, self.__stdin_cb)
        while True:
            try:
                super(IBusModeMainLoop, self).run()
            except:
                import traceback
                print '(error "%s")'%traceback.format_exc().replace('"', '\\"')
            else:
                break
        for ic in imcontexts:
            if ic:
                ic.destroy()


if __name__ == "__main__":

    mainloop = IBusModeMainLoop(bus)
    mainloop.run()


ibus-mode-agent.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

# ibus-mode-agent.py --- helper program of IBus client for GNU Emacs
# Copyright (c) 2009 and onwards, S. Irie

# Author: S. Irie
# Maintainer: S. Irie
# Version: 0.0.2.32

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Commentary:

# IBus is a new input method framework under active development
# which is designed to overcome the limitations of SCIM.

# IBus uses D-Bus protocol for communication between the ibus-daemon
# and clients (engines, panel, config tools). Since the components
# run in separate processes there is enhanced modularity and stability.
# Client processes can be loaded, started and stopped independently.
# IBus supports Gtk2 and XIM, and has input method engines for anthy,
# chewing, hangul, m17n, pinyin, rawcode, and large tables. Engines
# and clients can be written in any language with a dbus binding.

# This program is IBus client for GNU Emacs. It is, however,
# not part of official IBus project.

# ChangeLog:

# 2010-05-09
#        * Version 0.0.2
# 2010-04-12
#        * Version 0.0.1
#        * Initial experimental version

# Code:

import sys
import glib
import re

import ibus
from ibus import keysyms
from ibus import modifier

try:
    import Xlib.display
except ImportError:
    print "python-xlib library is required"
    exit(1)

display = Xlib.display.Display()

########################################################################
# Connect to IBus daemon
########################################################################

try:
    bus = ibus.Bus()
except TypeError:
    import os
    import time
    if os.spawnlp(os.P_WAIT, "ibus-daemon", "ibus-daemon", "-d") == 0:
        for i in range(10):
            time.sleep(0.5)
            try:
                bus = ibus.Bus()
                break
            except TypeError:
                pass
        else:
            print '(error "Failed to connect with ibus-daemon")'
            exit(1)
    else:
        print '(error "Failed to launch ibus-daemon")'
        exit(1)

########################################################################
# Miscellaneous functions
########################################################################

def lisp_boolean(boolean):
    return "t" if boolean else "nil"

escape_regexp = re.compile(ur'(["\\])')

def escape_string(string):
    return escape_regexp.sub(ur'\\\1', string)

########################################################################
# Input Context
########################################################################

class IBusModeIMContext(ibus.InputContext):

    def __init__(self):
        self.__path = bus.create_input_context("IBusModeIMContext")
        super(IBusModeIMContext, self).__init__(bus, self.__path, True)

        self.id_no = 0
        self.lookup_table = None

        self.connect('commit-text', self.__commit_text_cb)
        self.connect('update-preedit-text', self.__update_preedit_text_cb)
        self.connect('show-preedit-text', self.__show_preedit_text_cb)
        self.connect('hide-preedit-text', self.__hide_preedit_text_cb)
        self.connect('update-auxiliary-text', self.__update_auxiliary_text_cb)
        self.connect('show-auxiliary-text', self.__show_auxiliary_text_cb)
        self.connect('hide-auxiliary-text', self.__hide_auxiliary_text_cb)
        self.connect('update-lookup-table', self.__update_lookup_table_cb)
        self.connect('show-lookup-table', self.__show_lookup_table_cb)
        self.connect('hide-lookup-table', self.__hide_lookup_table_cb)
        self.connect('page-up-lookup-table', self.__page_up_lookup_table_cb)
        self.connect('page-down-lookup-table', self.__page_down_lookup_table_cb)
        self.connect('cursor-up-lookup-table', self.__cursor_up_lookup_table_cb)
        self.connect('cursor-down-lookup-table', self.__cursor_down_lookup_table_cb)
        self.connect('enabled', self.__enabled_cb)
        self.connect('disabled', self.__disabled_cb)
        try:
            self.connect('forward-key-event', self.__forward_key_event_cb)
        except TypeError:
            pass
        try:
            self.connect('delete-surrounding-text', self.__delete_surrounding_text_cb)
        except TypeError:
            pass

    # Callbacks

    def __commit_text_cb(self, ic, text):
        print '(ibus-commit-text-cb %d "%s")'% \
            (ic.id_no, escape_string(text.text).encode("utf-8"))

    def __update_preedit_text_cb(self, ic, text, cursor_pos, visible):
        attrs = ['%s %d %d %d'%
                 (["nil", "'underline", "'foreground", "'background"][attr.type],
                  attr.value & 0xffffff, attr.start_index, attr.end_index)
                 for attr in text.attributes]
        print '(ibus-update-preedit-text-cb %d "%s" %d %s %s)'% \
            (ic.id_no, escape_string(text.text).encode("utf-8"),
             cursor_pos, lisp_boolean(visible), ' '.join(attrs))

    def __show_preedit_text_cb(self, ic):
        print '(ibus-show-preedit-text-cb %d)'%(ic.id_no)

    def __hide_preedit_text_cb(self, ic):
        print '(ibus-hide-preedit-text-cb %d)'%(ic.id_no)

    def __update_auxiliary_text_cb(self, ic, text, visible):
        print '(ibus-update-auxiliary-text-cb %d "%s" %s)'% \
            (ic.id_no, escape_string(text.text).encode("utf-8"),
             lisp_boolean(visible))

    def __show_auxiliary_text_cb(self, ic):
        print '(ibus-show-auxiliary-text-cb %d)'%(ic.id_no)

    def __hide_auxiliary_text_cb(self, ic):
        print '(ibus-hide-auxiliary-text-cb %d)'%(ic.id_no)

    def __update_lookup_table_cb(self, ic, lookup_table, visible):
        ic.lookup_table = lookup_table
        if visible:
            self.__show_lookup_table_cb(ic)
        else:
            self.__hide_lookup_table_cb(ic)

    def __show_lookup_table_cb(self, ic):
        print "(ibus-show-lookup-table-cb %d '(%s) %s)"% \
            (ic.id_no, escape_string(
                " ".join(map(lambda item : '"%s"'%item.text,
                             ic.lookup_table.get_candidates_in_current_page())
                         )).encode("utf-8"),
             ic.lookup_table.get_cursor_pos_in_current_page())

    def __hide_lookup_table_cb(self, ic):
        print '(ibus-hide-lookup-table-cb %d)'%(ic.id_no)

    def __page_up_lookup_table_cb(self, ic):
        print '(ibus-log "page up lookup table")'

    def __page_down_lookup_table_cb(self, ic):
        print '(ibus-log "page down lookup table")'

    def __cursor_up_lookup_table_cb(self, ic):
        print '(ibus-log "cursor up lookup table")'

    def __cursor_down_lookup_table_cb(self, ic):
        print '(ibus-log "cursor down lookup table")'

    def __enabled_cb(self, ic):
        print '(ibus-status-changed-cb %d "%s")'%(ic.id_no, ic.get_engine().name)

    def __disabled_cb(self, ic):
        print '(ibus-status-changed-cb %d nil)'%ic.id_no

    def __forward_key_event_cb(self, ic, keyval, keycode, modifiers):
        print '(ibus-forward-key-event-cb %d %d %d %s)'% \
            (ic.id_no, keyval, modifiers & ~modifier.RELEASE_MASK,
             lisp_boolean(modifiers & modifier.RELEASE_MASK == 0))

    def __delete_surrounding_text_cb(self, ic, offset, n_chars):
        print '(ibus-delete-surrounding-text-cb %d %d %d)'% \
            (ic.id_no, offset, n_chars)

########################################################################
# Process methods from client
########################################################################

imcontexts = []

def create_imcontext():
    ic = IBusModeIMContext()
    imcontexts.append(ic)
    ic.id_no = len(imcontexts)-1
    ic.set_capabilities(int('101001',2))
    print '(ibus-create-imcontext-cb %d)'%ic.id_no

def destroy_imcontext(id_no):
    imcontexts[id_no].destroy()
    imcontexts[id_no] = None

def process_key_event(id_no, keyval, modmask, backslash, pressed):
    if backslash:
        keycode = (display.keysym_to_keycode(backslash), 0)
    else:
        keycode = display.keysym_to_keycodes(keyval)[0]
        if keycode[1] & 1 != 0:
            modmask |= modifier.SHIFT_MASK
    if not pressed:
        modmask |= modifier.RELEASE_MASK
    handled = imcontexts[id_no].process_key_event(keyval, keycode[0] - 8, modmask)
    print '(ibus-process-key-event-cb %d %s)'%(id_no, lisp_boolean(handled))

def set_cursor_location(id_no, x, y, w, h):
    imcontexts[id_no].set_cursor_location(x, y, w, h)

def focus_in(id_no):
    imcontexts[id_no].focus_in()

def focus_out(id_no):
    imcontexts[id_no].focus_out()
    print '()' # Dummy response

def reset(id_no):
    imcontexts[id_no].reset()

def enable(id_no):
    imcontexts[id_no].enable()

def disable(id_no):
    imcontexts[id_no].disable()

def set_engine(id_no, name):
    for engine in bus.list_active_engines():
        if name == '%s'%engine.name:
            imcontexts[id_no].set_engine(engine)
            break
    else:
        enable(id_no)

########################################################################
# Experiment
########################################################################

def get_engine(id_no):
    engine = imcontexts[id_no].get_engine()
    print '(ibus-log "engine: %s")'%(engine.name if engine else 'none')

def list_active_engines():
    print '(ibus-log "active engines: %s")'% \
        ' '.join('%s'%i.name for i in bus.list_active_engines())

def delete_surrounding_text(id_no, offset, n_chars):
    print '(ibus-delete-surrounding-text-cb %d %d %d)'% \
        (id_no, offset, n_chars)

def forward_key_event(id_no, keyval, keycode, modifiers):
    print '(ibus-forward-key-event-cb %d %d %d %s)'% \
        (id_no, keyval, modifiers & ~modifier.RELEASE_MASK,
         lisp_boolean(modifiers & modifier.RELEASE_MASK == 0))

########################################################################
# Main loop
########################################################################

class IBusModeMainLoop(glib.MainLoop):

    def __init__(self, bus):
        super(IBusModeMainLoop, self).__init__()
        bus.connect("disconnected", self.__disconnected_cb)

    def __disconnected_cb(self, *args):
        print '(ibus-log "disconnected")'
        exit()

    def __start_cb(self):
        print '(setq started t)'
        print 'Agent successfully started for display "%s"'% \
            display.get_display_name()
        return False

    def __stdin_cb(self, fd, condition):
        try:
            exec sys.stdin.readline()
        except:
            import traceback
            print '(error "%s")'%traceback.format_exc().replace('"', '\\"')
        return True

    def run(self):
        glib.idle_add(self.__start_cb)
        glib.io_add_watch(0, glib.IO_IN, self.__stdin_cb)
        while True:
            try:
                super(IBusModeMainLoop, self).run()
            except:
                import traceback
                print '(error "%s")'%traceback.format_exc().replace('"', '\\"')
            else:
                break
        for ic in imcontexts:
            if ic:
                ic.destroy()


if __name__ == "__main__":

    mainloop = IBusModeMainLoop(bus)
    mainloop.run()
